Allow `SSL_CTX` to use the new multiple certificate slots internally

samuel40791765 opened the pull request:
### Issues:
Addresses `CryptoAlg-1844`

### Description of changes: 
This builds upon the new framework in https://github.com/aws/aws-lc/pull/1086 to use the new multiple certificate slots available. The SSL connection should use the corresponding index of the certificate and pkey type assigned. As of now, the connection will directly use the last certificate that was set for it's connection. This corresponds with our original behavior before we reintroduced the multiple certificate slots framework, where we only maintained one `x509_leaf` and one `pkey per `CERT`.`
A subsequent PR will be introduced to allow the user to manually select which certificate slots they wish to use.

### Call-outs:
These functions change the certificate slot index:
* `SSL_CTX_use_certificate_ASN1` / `SSL_use_certificate_ASN1`/ `SSL_CTX_use_certificate` / `SSL_use_certificate`  (functions that call `ssl_set_cert` internally)
* `SSL_use_*_PrivateKey_*` (functions that call `ssl_set_pkey` internally)

The naming for the list of functions are pretty straightforward. Users setting these functions would have expected `SSL_CTX` to use the new set certificate/pkey, so it's natural that we switch over to that slot. Despite it seeming straightforward, definately let me know if I should add more documentation regarding this anyways

`SSL_CTX_select_current_cert` is a bit more interesting since it selects an existing cert in `SSL_CTX` to use. This will be implemented in one of the next PRs.
* https://github.com/aws/aws-lc/commit/44628096232d22160840b2fa2ea8438343879f6f

### Testing:
I've added a new test that sets up a connection with different certificate types and checks if the correct slot is being used.
`ED25519` is incompatible with older versions of TLS, so those are skipped.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

justsmth commented on ssl/ssl_lib.cc:
```diff
@@ -2476,16 +2485,21 @@ EVP_PKEY *SSL_get_privatekey(const SSL *ssl) {
   }
-  if (ssl->config->cert != NULL) {
-    return ssl->config->cert->privatekey.get();
+  if (ssl->config->cert != NULL &&
+      ssl_cert_check_cert_private_keys_usage(ssl->config->cert.get())) {
```
If the null check is added to `ssl_cert_check_cert_private_keys_usage`, then it's not needed here.

justsmth commented on ssl/ssl_privkey.cc:
```diff
@@ -665,7 +691,7 @@ static bool parse_sigalg_pairs(Array<uint16_t> *out, const int *values,
 
   for (size_t i = 0; i < num_values; i += 2) {
     const int hash_nid = values[i];
-    const int pkey_type = values[i+1];
+    const int pkey_type = values[i + 1];
```
Nit: For safety, the loop should terminate if  `i+1 >= num_values`
```
for (size_t i = 0; i < num_values - 1; i += 2) {
```

samuel40791765 commented on ssl/ssl_lib.cc:
```diff
@@ -1757,16 +1757,25 @@ int SSL_has_pending(const SSL *ssl) {
 }
 
 int SSL_check_private_key(const SSL *ssl) {
-  if (!ssl->config) {
```
* These were probably left over from OpenSSL that hadn't been refactored. OpenSSL code has a really bad habit of not doing proper null checks. It might be worth cleaning these up all at once in another SIM/PR.
* The behavior is well documented in `SSL_set_shutdown` so I believe its intentional. `SSL_CTX_set_quiet_shutdown` is encouraged instead.

samuel40791765 commented on ssl/ssl_cert.cc:
```diff
@@ -334,12 +334,29 @@ static int cert_set_chain_and_key(
   }
 
-  CERT_PKEY &cert_pkey = cert->cert_private_keys[cert->cert_private_key_idx];
+  int slot_index = ssl_get_certificate_slot_index_from_buffer(buffer.get());
+  if (slot_index < 0) {
+    return false;
+  }
+  CERT_PKEY &cert_pkey = cert->cert_private_keys[slot_index];
 
   switch (
       check_leaf_cert_and_privkey(buffer.get(), cert_pkey.privatekey.get())) {
```
Great point, I hadn't looked quite into that since it require some additional logic upon the stuff that's already here, but I'll work on it :).

samuel40791765 commented:
While working on Justin's PR comments, I noticed that `SSL_CTX_set_chain_and_key` (custom BoringSSL only function) also sets up the certificate and private key for use. I fixed it and added a test for it while I was at it.

justsmth commented on ssl/ssl_cert.cc:
```diff
@@ -250,45 +250,52 @@ enum leaf_cert_and_privkey_result_t {
       // Sanity-check that the private key and the certificate match.
-      !ssl_compare_public_and_private_key(pubkey.get(), privkey)) {
+      !ssl_compare_public_and_private_key(pubkey, privkey)) {
     ERR_clear_error();
```
Is there a reason we "clear" errors when the keys mismatch?
